apps/agreement/contracts/Agreement.sol

```diff
@@ -0,0 +1,1350 @@
+    * @param _actionId Identification number of the action to be queried
+    * @param _action Action instance to be queried
+    * @return True if the action can proceed, false otherwise
+    */
+    function _canProceed(uint256 _actionId, Action storage _action) internal view returns (bool) {
+        // If the action was already closed, return false
+        if (_action.closed) {
+            return false;
+        }
+
+        uint256 challengeId = _action.currentChallengeId;
+        Challenge storage challenge = challenges[challengeId];
+
+        // If the action was not challenged, return true
+        if (!_existChallenge(_actionId, challengeId, challenge)) {
+            return true;
+        }
+
+        // If the action was challenged but ruled in favour of the submitter
+        // (dispute rejected by arbitrator) or voided, return true
+        ChallengeState state = challenge.state;
+        return state == ChallengeState.Rejected || state == ChallengeState.Voided;
+    }
+
+    /**
+    * @dev Tell whether an action can be settled
+    * @param _actionId Identification number of the action to be queried
+    * @param _challenge Current challenge instance for the action
+    * @return True if the action can be settled, false otherwise
+    */
+    function _canSettle(uint256 _actionId, Challenge storage _challenge) internal view returns (bool) {
+        return _isWaitingChallengeAnswer(_actionId, _challenge);
+    }
+
+    /**
+    * @dev Tell whether an action can be disputed
+    * @param _actionId Identification number of the action to be queried
+    * @param _challenge Current challenge instance for the action
+    * @return True if the action can be disputed, false otherwise
+    */
+    function _canDispute(uint256 _actionId, Challenge storage _challenge) internal view returns (bool) {
+        if (!_isWaitingChallengeAnswer(_actionId, _challenge)) {
+            return false;
+        }
+
+        return _challenge.endDate > getTimestamp64();
+    }
+
+    /**
+    * @dev Tell whether an action settlement can be claimed
+    * @param _actionId Identification number of the action to be queried
+    * @param _challenge Current challenge instance for the action
+    * @return True if the action settlement can be claimed, false otherwise
+    */
+    function _canClaimSettlement(uint256 _actionId, Challenge storage _challenge) internal view returns (bool) {
+        if (!_isWaitingChallengeAnswer(_actionId, _challenge)) {
+            return false;
+        }
+
+        return getTimestamp64() >= _challenge.endDate;
+    }
+
+    /**
+    * @dev Tell whether an action's dispute can be ruled
+    * @param _actionId Identification number of the action to be queried
+    * @param _challenge Current challenge instance for the action
+    * @return True if the action's dispute can be ruled, false otherwise
+    */
+    function _canRuleDispute(uint256 _actionId, Challenge storage _challenge) internal view returns (bool) {
```

Comment:
Not sure we really need this indirection, maybe we can just use `_isDisputed()` directly?